src/dbnode/client/replicated_session.go

```diff
@@ -0,0 +1,301 @@
+type replicatedSession struct {
+	session       clientSession
+	asyncSessions []clientSession
+	newSessionFn  newSessionFn
+	workerPool    m3sync.WorkerPool
+	scope         tally.Scope
+	log           *zap.Logger
+	metrics       replicatedSessionMetrics
+	outCh         chan error
+}
+
+type replicatedSessionMetrics struct {
+	replicateExecuted    tally.Counter
+	replicateNotExecuted tally.Counter
+	replicateError       tally.Counter
+}
+
+func newReplicatedSessionMetrics(scope tally.Scope) replicatedSessionMetrics {
+	return replicatedSessionMetrics{
+		replicateExecuted:    scope.Counter("replicate.executed"),
+		replicateNotExecuted: scope.Counter("replicate.not-executed"),
+		replicateError:       scope.Counter("replicate.error"),
+	}
+}
+
+// Ensure replicatedSession implements the clientSession interface.
+var _ clientSession = (*replicatedSession)(nil)
+
+type replicatedSessionOption func(*replicatedSession)
+
+func withNewSessionFn(fn newSessionFn) replicatedSessionOption {
+	return func(session *replicatedSession) {
+		session.newSessionFn = fn
+	}
+}
+
+func newReplicatedSession(opts ReplicatedOptions, options ...replicatedSessionOption) (clientSession, error) {
+	// TODO(srobb): Replace with PooledWorkerPool once it has a GoIfAvailable method
+	workerPool := m3sync.NewWorkerPool(maxReplicationConcurrency)
+	workerPool.Init()
+
+	scope := opts.Options().InstrumentOptions().MetricsScope()
+
+	session := replicatedSession{
+		newSessionFn: newSession,
+		workerPool:   workerPool,
+		scope:        scope,
+		log:          opts.Options().InstrumentOptions().Logger(),
+		metrics:      newReplicatedSessionMetrics(scope),
+		outCh:        make(chan error),
+	}
+
+	// Apply options
+	for _, option := range options {
+		option(&session)
+	}
+
+	if err := session.setSession(opts.Options()); err != nil {
+		return nil, err
+	}
+	if err := session.setAsyncSessions(opts.OptionsForAsyncClusters()); err != nil {
+		return nil, err
+	}
+
+	return &session, nil
+}
+
+type writeFunc func(Session) error
+
+func (s *replicatedSession) setSession(opts Options) error {
+	if opts.TopologyInitializer() == nil {
+		return nil
+	}
+
+	session, err := s.newSessionFn(opts)
+	if err != nil {
+		return err
+	}
+	s.session = session
+	return nil
+}
+
+func (s *replicatedSession) setAsyncSessions(opts []Options) error {
+	sessions := make([]clientSession, 0, len(opts))
+	for _, oo := range opts {
+		session, err := s.newSessionFn(oo)
+		if err != nil {
+			return err
+		}
+		sessions = append(sessions, session)
+	}
+	s.asyncSessions = sessions
+	return nil
+}
+
+func (s replicatedSession) replicate(fn writeFunc) error {
+	for _, asyncSession := range s.asyncSessions {
+		asyncSession := asyncSession // capture var
+		if s.workerPool.GoIfAvailable(func() {
+			err := fn(asyncSession)
+			if err != nil {
+				s.metrics.replicateError.Inc(1)
+				s.log.Error("could not replicate write: %v", zap.Error(err))
+			}
+			s.outCh <- err
```

Comment:
Hm, don't you want to only write errors to this channel?  If it's only used for tests you may want to only allocate the channel when using for testing and do a `if.outCh != nil { ... }` check so it's not used for non-tests.

Also, this is going to block forever if no one is reading from the channel.

Best to make this a buffered channel (so some fixed size), then send on it in a way that it won't block the goroutine forever:
```go
select {
case s.outCh <- err:
default:
  // Out channel is full
}
```